### ISSUE TITLE
Inconsistent appium behaviour across system

### ISSUE THREAD

**Issue by hellozyemlya** [AvaloniaUI/Avalonia#10437]
**Is your feature request related to a problem? Please describe.**
Problem is that winappdriver and mac2 appium drivers producing very different results:

- mac focused on XCUIElement element and provides different functionality from winappdriver
- different xpaths that can't be used for cross-platform testing
- lot of bugs in mac2 driver(as well as winappdriver get a little bit outdated - see https://github.com/appium/appium-windows-driver/pull/162 for reference)

This makes automation tests creation process a little bit painful.

**Describe the solution you'd like**
Roll out own appium avalonia driver that will heavily rely on avalonia internals and expose them to appium in cross-platform way:

- follow internal avalonia structure and report it to appium
  - expose real avalonia class names as tags
  - expose classes on all operation systems(mac2 does not expose classes of elements)
  - etc.
- reuse winappdriver and XCUIElement-based framework for low level clicking and keystrokes(and respective mobile platforms alternatives)

**Comment by grokys** [comment]
Appium is quite frankly a shitshow on desktop. WinAppDriver is stuck with Appium 1 and MS have abandoned it, and the mac2 driver development is all happening for Appium 2 so we can't even get fixes for it.

Given infinite time I'd love to write our own Appium drivers but unfortunately we've not got infinite time ;) It would make a great community project though (hint hint)!

**Comment by hellozyemlya** [comment]
Worth to say, several bug-reports for mac2 were fixed last days. At least it does not fails on every simplest actions anymore :)

But anyway, is there some coordination how this integration testing stuff goes in avalonia?

**Comment by grokys** [comment]
> Worth to say, several bug-reports for mac2 were fixed last days. At least it does not fails on every simplest actions anymore :)

Yep, but if I understand correctly, will only be available when targeting Appium 2. Which we can't do because of WinAppDriver.

> But anyway, is there some coordination how this integration testing stuff goes in avalonia?

Not sure what you mean? IMO what is really needed as a start is a non-closed source, actively maintained Windows Appium driver.

**Comment by hellozyemlya** [comment]
But how opensource winappdriver implementation will help to resolve differences between mac and windows implementation of all that automation stuff?

**Comment by maxkatz6** [comment]
> This makes automation tests creation process a little bit painful.

It is. That's why I would recommend headless testing when it's applicable.

**Comment by hellozyemlya** [comment]
@maxkatz6 unfortunately headless testing will not cover any platform specific input or visualization bugs. Also headless testing will make QA teams and managers unhappy, since it will be harder to record fancy videos and make screenshots.

btw, where it possible to find headless tests example? Is they available somewhere in sources?

**Comment by grokys** [comment]
> But how opensource winappdriver implementation will help to resolve differences between mac and windows implementation of all that automation stuff?

From what I've seen, WinAppDriver often does stuff it's own way which is different from the other Appium drivers. An open-source Windows Appium driver would need to work together with the rest of the Appium ecosystem to try to unify behavior across platforms and make desktop testing a first-class citizen.

But this is an issue for the Appium ecosystem, it's not really anything to do with Avalonia.
